### PR TITLE
Added rest layer changes for List all PITs and PIT segments

### PR DESCRIPTION
**Pull request by bharath-techie**
### Description
The changes contain rest layer changes for list all PITs and PIT segments APIs.

### Issues Resolved
https://github.com/opensearch-project/OpenSearch/issues/1147
https://github.com/opensearch-project/OpenSearch/issues/3959

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/rest/action/cat/RestPitSegmentsAction.java
Comment (bharath-techie) [inline server/src/main/java/org/opensearch/rest/action/cat/RestPitSegmentsAction.java:67]: Because we need cluster state call within the rest layer, can't avoid this. Which means, PIT segments API will also need cluster state permissions to work.

File: client/rest-high-level/src/main/java/org/opensearch/client/RestHighLevelClient.java
Comment (pranikum) [inline client/rest-high-level/src/main/java/org/opensearch/client/RestHighLevelClient.java:66]: Any reason for this addition.

File: server/src/main/java/org/opensearch/action/search/GetAllPitNodesResponse.java
Comment (pranikum) [inline server/src/main/java/org/opensearch/action/search/GetAllPitNodesResponse.java:71]: ```getAllPitNodeResponse```. -> ```responses``` or``` responseList```

**Review by pranikum — APPROVED** [review]

**Comment by bharath-techie** [comment]
@reta @Bukhtawar can you please review this ? This is the final major PR for point in time changes. We need to check this in before code freeze (sep 7th)

File: server/src/main/java/org/opensearch/action/search/ListPitInfo.java
Comment (reta) [inline server/src/main/java/org/opensearch/action/search/ListPitInfo.java:65]: Please use CREATION_TIME, PIT_ID and KEEP_ALIVE here (this is one of the reasons to have the static:

```
    static {
        PARSER.declareString(constructorArg(), PIT_ID);
        PARSER.declareLong(constructorArg(), CREATION_TIME);
        PARSER.declareLong(constructorArg(), KEEP_ALIVE);
    }

``

File: server/src/main/java/org/opensearch/rest/action/cat/RestPitSegmentsAction.java
Comment (reta) [reply server/src/main/java/org/opensearch/rest/action/cat/RestPitSegmentsAction.java:70]: Why `PitSegmentsAction` needs cluster state permission? The `/_cat/...` certainly does, but not `PitSegmentsAction` I believe

File: server/src/main/java/org/opensearch/rest/action/cat/RestPitSegmentsAction.java
Comment (reta) [inline server/src/main/java/org/opensearch/rest/action/cat/RestPitSegmentsAction.java:42]: I believe one important piece of information (lost in `TransportPitSegmentsAction`) it the `pit_id`s the segment belongs to, what do you think?

File: server/src/main/java/org/opensearch/rest/action/cat/RestPitSegmentsAction.java
Comment (bharath-techie) [reply server/src/main/java/org/opensearch/rest/action/cat/RestPitSegmentsAction.java:70]: Since the cluster state is used in the table logic, we are making the cluster state api call here which makes it a dependency permission

File: server/src/main/java/org/opensearch/rest/action/search/RestGetAllPitsAction.java
Comment (bharath-techie) [reply server/src/main/java/org/opensearch/rest/action/search/RestGetAllPitsAction.java:60]: I agree, I too was thinking what to do in case of failures.
For delete all, if we don't find any pits to delete, we return 404,so thought of doing the same for list all pits.

File: server/src/main/java/org/opensearch/rest/action/cat/RestPitSegmentsAction.java
Comment (reta) [reply server/src/main/java/org/opensearch/rest/action/cat/RestPitSegmentsAction.java:70]: That's the `/_cat/...` API

Comment (bharath-techie) [reply server/src/main/java/org/opensearch/rest/action/cat/RestPitSegmentsAction.java:70]: Yeah only for` /_cat/...` API i meant that there is dependency. SegmentsAction by itself is not dependant on this.

File: server/src/main/java/org/opensearch/rest/action/search/RestGetAllPitsAction.java
Comment (bharath-techie) [reply server/src/main/java/org/opensearch/rest/action/search/RestGetAllPitsAction.java:62]: I'm throwing 500 for failures, does that sound okay ? can't think of anything else. please suggest

**Comment by bharath-techie** [comment]
@reta @Bukhtawar please review this PR. This PR is planned for 2.3

File: server/src/main/java/org/opensearch/action/search/ListPitInfo.java
Comment (reta) [inline server/src/main/java/org/opensearch/action/search/ListPitInfo.java:68]: Why do you need to recreate `ParseField` again since you already have them?

``
PARSER.declareString(constructorArg(), PIT_ID);
PARSER.declareString(constructorArg(), CREATION_TIME);
PARSER.declareString(constructorArg(), KEEP_ALIVE);
```

File: server/src/main/java/org/opensearch/client/ClusterAdminClient.java
Comment (reta) [inline server/src/main/java/org/opensearch/client/ClusterAdminClient.java:345]: I am confused now: we have non-admin `GetAllPitsAction` which supposed to list PITs by using `NodesGetAllPitsAction` (the `NodesGetAllPitsAction` was not supposed to be used directly initially).The `NodesGetAllPitsAction` is now used directly, so do we need `GetAllPitsAction`? I suspect the answer is "no"

Comment (bharath-techie) [reply server/src/main/java/org/opensearch/client/ClusterAdminClient.java:345]: Its a dummy action that is used to filter the permitted PITs.

So `NodesGetAllPitAction` -----makes a transport call to `GetAllPitsAction` ---> Which security plugin interjects and **replaces the list of PITs with list of permitted PITs**

This works because `GetAllPitsAction` is of type `indices:data/read` action, so PITs will be filtered based on user index permissions.

Comment (bharath-techie) [reply server/src/main/java/org/opensearch/client/ClusterAdminClient.java:345]: So,
1. Rest layer calls `NodesGetAllPitAction` which gets all PITs from cluster
2. `NodesGetAllPitAction` makes a call to `GetAllPitsAction` filled with a request that has all PITs 
3. Security plugin intercepts `GetAllPitsAction` to filter all the PITs based on indices permission
4. `GetAllPitsAction` will just send back the response.

Comment (reta) [reply server/src/main/java/org/opensearch/client/ClusterAdminClient.java:345]: But user now must have the permission to call "cluster:admin/point_in_time/readall" as well?

Comment (bharath-techie) [reply server/src/main/java/org/opensearch/client/ClusterAdminClient.java:345]: yeah that is already discussed [here](https://github.com/opensearch-project/OpenSearch/pull/4064#discussion_r954885360)

But instead of using security plugin to call list all pits ( which again required the permission, because it was a transport call )  we wanted to eliminate usage of latch, we made this alternate approach

File: server/src/main/java/org/opensearch/action/search/CreatePitController.java
Comment (bharath-techie) [reply server/src/main/java/org/opensearch/action/search/CreatePitController.java:202]: Correct , throwing error and ending the operation if that's the case.

File: server/src/main/java/org/opensearch/client/ClusterAdminClient.java
Comment (reta) [reply server/src/main/java/org/opensearch/client/ClusterAdminClient.java:345]: I see now,  thanks @bharath-techie ... I think out of these 2 options (dummy action vs security plugin), the first one is preferrable

Comment (bharath-techie) [reply server/src/main/java/org/opensearch/client/ClusterAdminClient.java:345]: yes, we're going with dummy action since it has no thread blocks and works seemlessly

Comment (reta) [reply server/src/main/java/org/opensearch/client/ClusterAdminClient.java:345]: @peternied do you think we should be using `ActionFilter` instead of dummy action? The idea would be:
 - use `ActionFilter` to intercept `NodesGetAllPitsAction` actions 
 - wrap the action listener (so we could hoon to `onResponse`)
 - on response, filter out PITs using the `PitPrivilegesEvaluator` from https://github.com/opensearch-project/security/pull/2053?

Comment (Bukhtawar) [reply server/src/main/java/org/opensearch/client/ClusterAdminClient.java:345]: +1 on ActionFilter to avoid the dummy action but @reta do you think we can change this incrementally over the next minor version release

Comment (reta) [reply server/src/main/java/org/opensearch/client/ClusterAdminClient.java:345]: @Bukhtawar if we get it into 2.x, that would be breaking change :(

Comment (bharath-techie) [reply server/src/main/java/org/opensearch/client/ClusterAdminClient.java:345]: 1. As per this, we will remove `GetAllPitsAction` which is `indices:data/read`.
But `NodesGetAllPitsAction` will still be present and user will need `cluster:admin/point_in_time`. 
Are we okay with this ?

2. We need to modify / repeat a lot of logic in `pit filter` which is present in `security filter` like admin check ( which should pass through without any additional logic) and per request we again need to constuct user , security roles, dependent object which will be duplicated in `pit filter` from `security filter` . Main concern is some new logic if added to security filter should be also added to `pit filter` since it has dependant objects causing maintainability issues.

```
if(pitPrivilegesEvaluator.evaluate(request, clusterService, user, securityRoles,
                    action0, resolver, dcm.isDnfofForEmptyResultsEnabled(), presponse).isComplete()) {
```

Comment (Bukhtawar) [reply server/src/main/java/org/opensearch/client/ClusterAdminClient.java:345]: @reta with the current proposal we aren't getting rid of `cluster:admin` which is where this discussion started in the first place(we had a concern that there shouldn't be a cluster management permission for PIT). Please correct me if my understanding isn't correct, just catching up. 
Taking a step back do we really need ActionFilter changes to start with if we could live with the original PR where the only concern was cluster management as a proxy.

Comment (reta) [reply server/src/main/java/org/opensearch/client/ClusterAdminClient.java:345]: @Bukhtawar ideally, we want `GetAllPitsAction` with `indices:data/read` permissions only, the problematic implementation we had that was using transpost action call (with latches, etc) because of inabitily to access the listener. With `ActionFilter` we do have access to the listener over a specific action, so we should be able to cleanly filter the result out, at least it looks feasible, let me prototype this solution (will update shortly).

Comment (bharath-techie) [reply server/src/main/java/org/opensearch/client/ClusterAdminClient.java:345]:
```
package org.opensearch.security.filter;

import com.google.common.collect.ImmutableSet;
import org.opensearch.OpenSearchSecurityException;
import org.opensearch.action.ActionListener;
import org.opensearch.action.ActionRequest;
import org.opensearch.action.ActionResponse;
import org.opensearch.action.search.GetAllPitNodesRequest;
import org.opensearch.action.support.ActionFilter;
import org.opensearch.action.support.ActionFilterChain;
import org.opensearch.cluster.metadata.IndexNameExpressionResolver;
import org.opensearch.cluster.service.ClusterService;
import org.opensearch.common.transport.TransportAddress;
import org.opensearch.common.util.concurrent.ThreadContext;
import org.opensearch.rest.RestStatus;
import org.opensearch.security.auditlog.AuditLog;
import org.opensearch.security.auth.RolesInjector;
import org.opensearch.security.configuration.AdminDNs;
import org.opensearch.security.privileges.PitPrivilegesEvaluator;
import org.opensearch.security.privileges.PrivilegesEvaluator;
import org.opensearch.security.privileges.PrivilegesEvaluatorResponse;
import org.opensearch.security.privileges.PrivilegesInterceptor;
import org.opensearch.security.securityconf.ConfigModel;
import org.opensearch.security.securityconf.DynamicConfigModel;
import org.opensearch.security.securityconf.SecurityRoles;
import org.opensearch.security.support.ConfigConstants;
import org.opensearch.security.support.HeaderHelper;
import org.opensearch.security.user.User;
import org.opensearch.tasks.Task;
import org.opensearch.threadpool.ThreadPool;

import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

public class PitFilter  implements ActionFilter {

    private final ClusterService clusterService;

    private final IndexNameExpressionResolver resolver;

    private final AuditLog auditLog;
    private ThreadContext threadContext;

    private PrivilegesInterceptor privilegesInterceptor;

    private final AdminDNs adminDns;
    private final RolesInjector rolesInjector;

    private ConfigModel configModel;

    private DynamicConfigModel dcm;


    private PrivilegesEvaluator privilegesEvaluator;

    @Override
    public int order() {
        return 0;
    }

    public PitFilter(final ClusterService clusterService, final ThreadPool threadPool, final IndexNameExpressionResolver resolver,
                     AuditLog auditLog, final PrivilegesInterceptor privilegesInterceptor,  AdminDNs adminDns, PrivilegesEvaluator privilegesEvaluator, DynamicConfigModel dcm, ConfigModel configModel) {
        this.clusterService = clusterService;
        this.resolver = resolver;
        this.auditLog = auditLog;

        this.threadContext = threadPool.getThreadContext();
        this.privilegesInterceptor = privilegesInterceptor;

        this.adminDns = adminDns;
        this.rolesInjector = new RolesInjector(auditLog);
        this.privilegesEvaluator = privilegesEvaluator;
        this.configModel = configModel;
        this.dcm = dcm;
    }

    @Override
    public <Request extends ActionRequest, Response extends ActionResponse> void apply(Task task, String action,
                                                                                       Request request, ActionListener<Response> listener, ActionFilterChain<Request, Response> chain) {
        if(! (request instanceof GetAllPitNodesRequest)) {
            chain.proceed(task, action, request, listener);
            return;
        }

        User user = threadContext.getTransient(ConfigConstants.OPENDISTRO_SECURITY_USER);
        final boolean userIsAdmin = isUserAdmin(user, this.adminDns);
        final boolean confRequest = "true".equals(HeaderHelper.getSafeFromHeader(threadContext, ConfigConstants.OPENDISTRO_SECURITY_CONF_REQUEST_HEADER));


        if(userIsAdmin
                || confRequest){

            if(userIsAdmin && !confRequest) {
                auditLog.logGrantedPrivileges(action, request, task);
                auditLog.logIndexEvent(action, request, task);
            }

            chain.proceed(task, action, request, listener);
            return;
        }

        final PrivilegesEvaluatorResponse presponse = new PrivilegesEvaluatorResponse();

        final TransportAddress caller = threadContext.getTransient(ConfigConstants.OPENDISTRO_SECURITY_REMOTE_ADDRESS);
        final Set<String> injectedRoles = rolesInjector.injectUserAndRoles(request, action, task, threadContext);
        Set<String> mappedRoles = (injectedRoles == null) ? mapRoles(user, caller) : injectedRoles;
        final String injectedRolesValidationString = threadContext.getTransient(ConfigConstants.OPENDISTRO_SECURITY_INJECTED_ROLES_VALIDATION);
        if(injectedRolesValidationString != null) {
            HashSet<String> injectedRolesValidationSet = new HashSet<>(Arrays.asList(injectedRolesValidationString.split(",")));
            if(!mappedRoles.containsAll(injectedRolesValidationSet)) {
                presponse.allowed = false;
                presponse.missingSecurityRoles.addAll(injectedRolesValidationSet);
            }
            mappedRoles = ImmutableSet.copyOf(injectedRolesValidationSet);
        }
        presponse.resolvedSecurityRoles.addAll(mappedRoles);
        final SecurityRoles securityRoles = privilegesEvaluator.getSecurityRoles(mappedRoles);

        PitPrivilegesEvaluator pitPrivilegesEvaluator = new PitPrivilegesEvaluator();
        ActionListener wrappedActionListener = ActionListener.wrap( r-> {
            // check access for point in time requests
            PrivilegesEvaluatorResponse pres  = pitPrivilegesEvaluator.evaluate(request, clusterService, user, securityRoles,
                    action, resolver, dcm.isDnfofForEmptyResultsEnabled(), presponse);
            if (pres.isAllowed()) {


            }
            else {
                String err;
                if(!pres.getMissingSecurityRoles().isEmpty()) {
                    err = String.format("No mapping for %s on roles %s", user, pres.getMissingSecurityRoles());
                } else {
                    err = (injectedRoles != null) ?
                            String.format("no permissions for %s and associated roles %s", pres.getMissingPrivileges(), pres.getResolvedSecurityRoles()) :
                            String.format("no permissions for %s and %s", pres.getMissingPrivileges(), user);
                }
                listener.onFailure(new OpenSearchSecurityException(err, RestStatus.FORBIDDEN)));
            }
        }, e -> {
            listener.onFailure(e);
        });
        chain.proceed(task, action, request, listener);
    }

    private static boolean isUserAdmin(User user, final AdminDNs adminDns) {
        if (user != null && adminDns.isAdmin(user)) {
            return true;
        }

        return false;
    }

    public Set<String> mapRoles(final User user, final TransportAddress caller) {
        return this.configModel.mapSecurityRoles(user, caller);
    }
}

```

Comment (reta) [reply server/src/main/java/org/opensearch/client/ClusterAdminClient.java:345]: @bharath-techie I beliebe the answer is "yes", correct? Now, the repeating part from `SecurityFilter` could be removed I believe, we could embeed `ActionFilter`s inside `SecurityFilter` (like nested ones) and look after the necessary security checks look if there is more specifc `ActionFilter` for an action (GetAllPitNodesRequest), and if yes - proceed with that, if no - fallback to generic processing

Comment (bharath-techie) [reply server/src/main/java/org/opensearch/client/ClusterAdminClient.java:345]: @reta why can't i just wrap the listener in security filter itself ? why do we need a new filter. because anyways we need a type check and the new filter is going to be nested inside

Comment (bharath-techie) [reply server/src/main/java/org/opensearch/client/ClusterAdminClient.java:345]: Also would like @peternied 's thoughts on the new filter in security plugin.

Comment (reta) [reply server/src/main/java/org/opensearch/client/ClusterAdminClient.java:345]: > @reta why can't i just wrap the listener in security filter itself ? why do we need a new filter. because anyways we need a type check and the new filter is going to be nested inside

We certainly could, that would not scale well if we need to add more actions (and the code is cleaner this way).

Comment (peternied) [reply server/src/main/java/org/opensearch/client/ClusterAdminClient.java:345]: I'll defer to @reta's expertise in this area

Comment (bharath-techie) [reply server/src/main/java/org/opensearch/client/ClusterAdminClient.java:345]: https://github.com/opensearch-project/OpenSearch/pull/4388#discussion_r963995407 --- please check this out. The changes will be in security plugin , so your expertise is also required.

Comment (peternied) [reply server/src/main/java/org/opensearch/client/ClusterAdminClient.java:345]: If you'd like more input on the design as it pertains to the security plugin, please create a draft PR where you can callout the concerns or areas you'd like feedback on.  If we feel we shouldn't merge this change without updates to security updates, could a feature flag be used to keep PIT disabled by default until all the components are in place?

Comment (bharath-techie) [reply server/src/main/java/org/opensearch/client/ClusterAdminClient.java:345]: https://github.com/opensearch-project/security/issues/2087 --- created this so that we can conclude on the design and then go forward with implementation PR.

**Comment by reta** [comment]
Looks like we are getting there, thanks @bharath-techie !

File: server/src/main/java/org/opensearch/action/search/CreatePitController.java
Comment (reta) [inline server/src/main/java/org/opensearch/action/search/CreatePitController.java:217]: Please `node` ->  `entry.getValue().getNode()` since it is `null` at this point

File: server/src/main/java/org/opensearch/action/admin/indices/segments/PitSegmentsRequest.java
Comment (reta) [inline server/src/main/java/org/opensearch/action/admin/indices/segments/PitSegmentsRequest.java:103]:
```suggestion
                                throw new IllegalArgumentException("pit_id array element should only contain PIT identifiers");
```

File: server/src/main/java/org/opensearch/action/admin/indices/segments/PitSegmentsRequest.java
Comment (reta) [inline server/src/main/java/org/opensearch/action/admin/indices/segments/PitSegmentsRequest.java:109]:
```suggestion
                            throw new IllegalArgumentException("pit_id element should only contain PIT indentifier");
```

File: server/src/main/java/org/opensearch/rest/action/cat/RestPitSegmentsAction.java
Comment (reta) [reply server/src/main/java/org/opensearch/rest/action/cat/RestPitSegmentsAction.java:39]: @bharath-techie wdyt?

Comment (reta) [reply server/src/main/java/org/opensearch/rest/action/cat/RestPitSegmentsAction.java:39]: We could probably create a followup issue to work on that

Comment (bharath-techie) [reply server/src/main/java/org/opensearch/rest/action/cat/RestPitSegmentsAction.java:39]: Sorry missed this - will add an issue and work as a followup , we were thinking if we can group "index, shard, prirep, ip, id and pitId" and show results below for this group.

```
-- index, shard, prirep, ip, id , pitId
----- segment, generation, docs.count, docs.deleted, size, size.memory, commited, searchable, version, compound
```

**Pull request merged**
